bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -648,6 +650,64 @@ std::pair<PrePrepareMsg *, bool> ReplicaImp::buildPrePrepareMsgBatchByRequestsNu
   return buildPrePrepareMessageByRequestsNum(requiredRequestsNum);
 }
 
+bool ReplicaImp::tryToSendDataPrePrepareMsg() {
+  // no need to check for pre-requsites
+  removeDuplicatedRequestsFromRequestsQueue();
+  PrePrepareMsg *pp = nullptr;
+  bool isSent = false;
+
+  auto batchedReq = reqBatchingLogic_.batchRequests();
+  isSent = batchedReq.second;
+  if (isSent) {
+    pp = batchedReq.first;
+    batch_closed_on_logic_on_++;
+    accumulating_batch_time_.add(
+        std::chrono::duration_cast<std::chrono::microseconds>(getMonotonicTime() - time_to_collect_batch_).count());
+    accumulating_batch_avg_time_.Get().Set((uint64_t)accumulating_batch_time_.avg());
+    if (accumulating_batch_time_.numOfElements() == 1000) {
+      accumulating_batch_time_.reset();  // We reset the average on every 1000 samples
+    }
+    time_to_collect_batch_ = MinTime;
+  }
+
+  if (!pp) return isSent;
+  // check for data separation
+  if (pp->requestsSize() >= config_.threshBatchSizeForDataSeparation) {
+    auto new_pp = getConsensusPPFromDataPP(pp);
+    if (new_pp) {
+      // add time service req
+      if (config_.timeServiceEnabled) {
+        auto timeServiceMsg = time_service_manager_->createClientRequestMsg();
+        new_pp->addRequest(timeServiceMsg->body(), timeServiceMsg->size());
+      }
+      InternalMessage im = ConsensusOnlyPPMsg{new_pp};
+      // add a client request to it
```

Comment:
why do we want to add it as a client message?
How can we make sure that this client message won't go in a different PP (or even in a dataPP)?